tests/test_metaprogramming.py

```diff
@@ -0,0 +1,378 @@
+        with python:
+            a = 1
+            with exo:
+                b = a
+
+    c_file, _ = compile_procs_to_strings([foo], "test.h")
+    assert c_file == golden
+
+
+def test_scope_collision3():
+    with pytest.raises(
+        NameError,
+        match="free variable 'x' referenced before assignment in enclosing scope",
+    ):
+
+        @proc
+        def foo(a: i32, b: i32):
+            with python:
+                with exo:
+                    a = b * x
+                x = 1
+
+
+def test_type_quote_elision(golden):
+    T = "i8"
+
+    @proc
+    def foo(a: T, x: T[2]):
+        a += x[0]
+        a += x[1]
+
+    c_file, _ = compile_procs_to_strings([foo], "test.h")
+    assert c_file == golden
+
+
+def test_unquote_in_slice(golden):
+    @proc
+    def foo(a: [i8][2]):
+        a[0] += a[1]
+
+    @proc
+    def bar(a: i8[10, 10]):
+        with python:
+            x = 2
+            with exo:
+                for i in seq(0, 5):
+                    foo(a[i, {x} : {2 * x}])
+
+    c_file, _ = compile_procs_to_strings([foo, bar], "test.h")
+    assert c_file == golden
+
+
+def test_unquote_slice_object1(golden):
+    @proc
+    def foo(a: [i8][2]):
+        a[0] += a[1]
+
+    @proc
+    def bar(a: i8[10, 10]):
+        with python:
+            for s in [slice(1, 3), slice(5, 7), slice(2, 4)]:
+                with exo:
+                    for i in seq(0, 10):
+                        foo(a[i, s])
+
+    c_file, _ = compile_procs_to_strings([foo, bar], "test.h")
+    assert c_file == golden
+
+
+def test_unquote_slice_object2():
+    with pytest.raises(
+        ParseError, match="cannot perform windowing on left-hand-side of an assignment"
+    ):
+
+        @proc
+        def foo(a: i8[10, 10]):
+            with python:
+                for s in [slice(1, 3), slice(5, 7), slice(2, 4)]:
+                    with exo:
+                        for i in seq(0, 10):
+                            a[i, s] = 2
+
+
+def test_unquote_index_tuple(golden):
+    @proc
+    def foo(a: [i8][2, 2]):
+        a[0, 0] += a[0, 1]
+        a[1, 0] += a[1, 1]
+
+    @proc
+    def bar(a: i8[10, 10, 10]):
+        with python:
+
+            def get_index(i):
+                return slice(i, ~{i + 2}), slice(~{i + 1}, ~{i + 3})
+
+            with exo:
+                for i in seq(0, 7):
+                    foo(a[i, {get_index(i)}])
+
+    c_file, _ = compile_procs_to_strings([foo, bar], "test.h")
+    assert c_file == golden
+
+
+def test_unquote_err():
+    with pytest.raises(
+        ParseError, match="Unquote computation did not yield valid type"
+    ):
+        T = 1
+
+        @proc
+        def foo(a: T):
+            a += 1
+
+
+def test_quote_complex_expr(golden):
+    @proc
+    def foo(a: i32):
+        with python:
+
+            def bar(x):
+                return ~{x + 1}
+
+            with exo:
+                a = {bar(~{a + 1})}
+
+    c_file, _ = compile_procs_to_strings([foo], "test.h")
```

Comment:
Why are you compiling the proc into C for every test? I think you should also sanity-check the ExoIR printing.
Can modify the asserts to sth like:
```
assert str(foo) + c_file == golden
```

Reply:
Also, you know that you can update all the golden tests by running `pytest tests/test_metaprogramming.py --update-golden` , right?

Reply:
complete